github.go

```diff
@@ -883,3 +884,265 @@ func (c *Client) ListReposBylanguage(owner string, lang string) ([]*github.Repos
+			// but the starLowerBound did not get lower,
+			// that means there's more than 1K repos with that specific star count,
+			// but we can't get more than 1K;
+			// we can only retrieve the first 1K of repos with any specific star count.
+
+			// If starLowerBound is zero, it means we can't go any lower, and we're done:
+			if starLowerBound == 0 {
+				break GetterLoop
+			}
+
+			useStarBound = true
+			if starLowerBound == latestStarCount {
+				// For any given starLowerBound, we can only retrieve the first 1K repos.
+				// For this particular starLowerBound,
+				// there are more than 1K repos.
+
+				// Let's decrement starLowerBound by one.
+				// This will skip any repo with that specific star count beyond the initial 1K repos that we already got.
+				latestStarCount--
+			}
+			starLowerBound = latestStarCount
+			opt.Page = 1 // Restart
+			continue GetterLoop
+		}
+		opt.Page = resp.NextPage
+	}
+
+	return allRepos, nil
+}
+
+// SearchRepos will return a list of repos that match the provided query.
+// NOTE: the repo search API does not search inside the repo contents,
+// but only in its meta (repo name, description, etc.)
+// For more info about query syntax and parameters, see:
+// https://docs.github.com/en/free-pro-team@latest/github/searching-for-information-on-github/searching-for-repositories
+func (c *Client) SearchRepos(query string) ([]*github.Repository, error) {
+
+	client := c.client
+
+	opt := &github.SearchOptions{
+		ListOptions: github.ListOptions{PerPage: 100},
+	}
+	// get all pages of results
+	var allRepos []*github.Repository
+	for {
+		var repos *github.RepositoriesSearchResult
+		var resp *github.Response
+		errs := RetryExponentialBackoff(9999, time.Second, func() error {
+			var err error
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+			defer cancel()
+
+			repos, resp, err = client.Search.Repositories(ctx, query, opt)
+			if err != nil {
+				return fmt.Errorf("error while executing request: %s", err)
+			}
+			ResponseCallback(resp)
+			if handleRateLimitError(err, resp) {
+				return err
+			}
+
+			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNotFound && resp.StatusCode != http.StatusNoContent {
+				// TODO: catch rate limit error, and wait
+				return fmt.Errorf(
+					"status code is: %v (%s)",
+					resp.StatusCode,
+					resp.Status,
+				)
+			}
+			// nil on 200 and 404
+			return nil
+		})
+		if errs != nil && len(errs) > 0 {
+			return nil, errors.New(FormatErrorArray("", errs))
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			// TODO: catch rate limit error, and wait
+			return nil, ErrNotFound
+		}
+
+		for repIndex := range repos.Repositories {
+			allRepos = append(allRepos, &repos.Repositories[repIndex])
+		}
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+
+	return allRepos, nil
+}
+
+func (c *Client) SearchCode(query string) ([]*github.Repository, error) {
```

Comment:
Looks good.

I'd prefer leaving to extraction of the `Repository` to the caller of `SearchCode`.

```suggestion
func (c *Client) SearchCode(query string) ([]*github.CodeResult, error) {
```

Reply:
Should L.1100 be changed to `var allRepos []*github.CodeResult` then?

Reply:
Yep, and renamed to something like `allCodeResults`

Reply:
Okay, done.